### PR TITLE
Qutrits: Gell-Mann Observable

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
There are very few simulators available today that offer qudit simulation. One notable example is [Google's Cirq](https://quantumai.google/cirq/qudits), which includes support for qudit operations. However, the simulator is far from complete, and a key drawback is the lack of operations available, which creates a need for users to define operations themselves. Popular libraries such as Qiskit and QuTiP also include some qudit manipulation capabilities, but with few use cases. Beyond these libraries, qudit simulation solutions are quite limited. Moreover, qudit simulation is a feature that is already in demand, as seen in this [issue](https://github.com/PennyLaneAI/pennylane/issues/2190) on the PennyLane repository.

This PR builds upon the previous PR's that added qutrit functionality by adding the Gell-Mann matrices as qutrit observables.

**Description of the Change:**
* Added `GellMannObs` observable(s), which are the qutrit analog of the Pauli observables.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/2190

### CODE REVIEW

File: pennylane/ops/qutrit/observables.py [hunk 1]
```diff
@@ -14,13 +14,26 @@
 """
 This submodule contains the qutrit quantum observables.
 """
-from pennylane import numpy as np
+import numpy as np
```
Comment (Jaybsoni) [inline pennylane/ops/qutrit/observables.py:17]: Why was this swapped? PL numpy should just dispatch to numpy in the backend?

File: pennylane/ops/qutrit/observables.py [hunk 2]
```diff
@@ -133,3 +146,136 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+        index (int): The index of the Gell-Mann matrix to be used. Must be between 1
+            and 8 inclusive
+        wires (Sequence[int] or int): the wire(s) the observable acts on
+        do_queue (bool): Indicates whether the operator should be
+            immediately pushed into the Operator queue (optional)
+        id (str or None): String representing the operation (optional)
+    """
+    num_wires = 1
+    num_params = 1
+    """int: Number of trainable parameters the operator depends on"""
```
Comment (Jaybsoni) [inline pennylane/ops/qutrit/observables.py:184]: This parameter shouldn't be trainable correct ?

File: pennylane/ops/qutrit/observables.py [hunk 3]
```diff
@@ -14,13 +14,26 @@
...
+# Array containing all 8 Gell-Mann matrices
+gm_mats = np.zeros((8, 3, 3), dtype=np.complex128)
+gm_mats[0] = np.array([[0, 1, 0], [1, 0, 0], [0, 0, 0]])
+gm_mats[1] = np.array([[0, -1j, 0], [1j, 0, 0], [0, 0, 0]])
+gm_mats[2] = np.diag([1, -1, 0])
+gm_mats[3] = np.array([[0, 0, 1], [0, 0, 0], [1, 0, 0]])
+gm_mats[4] = np.array([[0, 0, -1j], [0, 0, 0], [1j, 0, 0]])
+gm_mats[5] = np.array([[0, 0, 0], [0, 0, 1], [0, 1, 0]])
+gm_mats[6] = np.array([[0, 0, 0], [0, 0, -1j], [0, 1j, 0]])
+gm_mats[7] = np.diag([1, 1, -2]) / np.sqrt(3)
```
Comment (Jaybsoni) [inline pennylane/ops/qutrit/observables.py:34]: Is this some well known order in literature? Or was it assigned from some formula? Maybe worth having the formula which takes index --> GM matrix in the docs? Although we have them explicitly listed so its not a major concern

File: pennylane/ops/qutrit/observables.py [hunk 4]
```diff
@@ -133,3 +146,136 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+
+        >>> qml.GellMannObs.compute_diagonalizing_gates(3, wires=[0, 1])
+        []
+        """
+        if index in {3, 8}:
+            return []
+
+        _, v = np.linalg.eigh(gm_mats[index - 1])
+        v = np.roll(v, 1, axis=1)
+        return [QutritUnitary(v.conj().T, wires=wires)]
```
Comment (Jaybsoni) [inline pennylane/ops/qutrit/observables.py:281]: A suggestion for performance, maybe these quantities could be saved and hard coded in as opposed to being computed each time.

File: pennylane/ops/qutrit/observables.py [hunk 5]
```diff
@@ -133,3 +146,136 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+            index (int): The index of the Gell-Mann matrix to be used. Must be between 1
+            and 8 inclusive
+            wires (Iterable[Any], Wires): wires that the operator acts on
+        Returns:
+            list[.Operator]: list of diagonalizing gates
+
+        **Example**
+
+        >>> qml.GellMannObs.compute_diagonalizing_gates(3, wires=[0, 1])
+        []
```
Comment (Jaybsoni) [inline pennylane/ops/qutrit/observables.py:274]: Maybe a non-empty example would be more useful? Or print the matrix before so the user sees it is diagonal to begin with?

File: tests/devices/test_default_qutrit.py [hunk 6]
```diff
@@ -405,14 +401,34 @@ class TestExpval:
                 0,
                 [[0, -1j, 0], [1j, 0, 0], [0, 0, 0]],
             ),
+            (qml.GellMannObs, [1, 0, 0], 0, 1),
+            (qml.GellMannObs, [0, 0, 1], 0, 1),
```
Comment (Jaybsoni) [inline tests/devices/test_default_qutrit.py:405]: These two entries differ in state, but match in expected output, just want to confirm this is as expected?

**Review by Jaybsoni — COMMENTED** [review]
Was able to give it a quick initial skim, looks good for the most part. Just a few questions / comments

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -73,6 +73,10 @@
 * Added the `qml.TAdd` and `qml.TSWAP` qutrit operations which are the ternary analogs of the CNOT and SWAP operations respectively.
   ([#2843](https://github.com/PennyLaneAI/pennylane/pull/2843))
 
+* Added the `qml.GellMannObs` qutrit observable, which is the ternary generalization of the Pauli observables. Users must include an index as the first
```
Comment (glassnotes) [inline doc/releases/changelog-dev.md:76]: Optional, but what about shortening this to just `GellMann`? None of the other gates / observables specify what they are in their name.

File: pennylane/ops/qutrit/observables.py (hunk 3 above)
Comment (glassnotes) [reply pennylane/ops/qutrit/observables.py:34]: It seems to be a fairly common ordering; I'm actually not sure if it's derived from anything. But yeah it wouldn't hurt to put a comment both here for developers, and then in the actual docstring as well.

File: pennylane/ops/qutrit/observables.py (hunk 1 above)
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:17]: Why was this changed?

File: pennylane/ops/qutrit/observables.py (hunk 2 above)
Comment (glassnotes) [reply pennylane/ops/qutrit/observables.py:184]: Yes that's right.

File: pennylane/ops/qutrit/observables.py [hunk 8]
```diff
@@ -133,3 +146,136 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+        \lambda_{4} = \left(\begin{array}{ccc} 0 & 0 & 1 \\ 0 & 0 & 0\\ 1 & 0 & 0\end{array}\right)\;\;\;\;\;\;\;\;\;\;
+        \displaystyle \lambda_{5} = \left(\begin{array}{ccc} 0 & 0 & -i \\ 0 & 0 & 0\\ i & 0 & 0\end{array}\right) \;\;\;\;\;\;\;\;\;\; \\
+        \lambda_{6} = \left(\begin{array}{ccc} 0 & 0 & 0 \\ 0 & 0 & 1\\ 0 & 1 & 0\end{array}\right)
+        \lambda_{7} = \left(\begin{array}{ccc} 0 & 0 & 0 \\ 0 & 0 & -i\\ 0 & i & 0\end{array}\right)\;\;\;\;\;\;\;\;\;\;
+        \lambda_{8} = \frac{1}{\sqrt{3}}\left(\begin{array}{ccc} 1 & 0 & 0 \\ 0 & 1 & 0\\ 0 & 0 & -2\end{array}\right)\\
+
+    **Details:**
+
+    * Number of wires: 1
+    * Number of parameters: 1
```
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:171]:
```suggestion
    * Number of parameters: 0
```
(See below?)

File: pennylane/ops/qutrit/observables.py [hunk 9]
```diff
@@ -133,3 +146,136 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+
+
+class GellMannObs(Observable):
+    r"""
+    The Gell-Mann observables for qutrits
+
+    The Gell-Mann matrices are a set of 8 linearly independent :math:`3x3` traceless, Hermitian matrices which
+    naturally generalize the Pauli matrices from :math:`SU(2)` to :math:`SU(3)`.
+
+    .. math::
```
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:158]: The rendering here still looks a little odd; maybe try using an `align` or table environment?

image

File: pennylane/ops/qutrit/observables.py [hunk 10]
```diff
@@ -133,3 +146,136 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+    """
+    num_wires = 1
+    num_params = 1
+    """int: Number of trainable parameters the operator depends on"""
+
+    def __init__(self, index, wires, do_queue=True, id=None):
+        if not isinstance(index, int) or index < 1 or index > 8:
+            raise ValueError("The index must be an integer between 1 and 8 inclusive.")
+
+        super().__init__(index, wires, do_queue=do_queue, id=id)
```
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:190]: Further to the point above about the parameter not being trainable: this is calling the constructor of `Operator`, and passing index in the location of `params` which is meant to be the trainable argument. It would be better to have the signature something like

```
GellMannObs(wires=0, index=1)
```
 to distinguish it from parametrized operations / observables that take a potentially trainable value as argument (like `Hermitian`).

For reference, check out how this is done for operations like `MultiControlledX`, which have a number of non-trainable parameters as keyword arguments.

File: pennylane/ops/qutrit/observables.py [hunk 11]
```diff
@@ -133,3 +146,136 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+        the sequence of diagonalizing gates implements the unitary :math:`U`.
+
+        The diagonalizing gates rotate the state into the eigenbasis
+        of the operator.
+
+        .. seealso:: :meth:`~.GellMannObs.diagonalizing_gates`.
+
+        Args:
+            index (int): The index of the Gell-Mann matrix to be used. Must be between 1
+            and 8 inclusive
```
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:266]:
```suggestion
            index (int): The index of the Gell-Mann matrix to be used. Must be between 1 and 8 inclusive
```
Applies to all instances of this docstring throughout the file. Either indent the wrapped part four spaces so it doesn't align with the names of the arguments, or just make this one line, since it's really not that long.

File: pennylane/ops/qutrit/observables.py (hunk 4 above)
Comment (glassnotes) [reply pennylane/ops/qutrit/observables.py:281]: Yeah since all the eigenvalues are the same too (except for \lambda_8) can just do something similar as in `compute_eigvals` I think?

File: tests/devices/test_default_qutrit.py [hunk 12]
```diff
@@ -797,15 +830,171 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
+            [
+                qml.QutritUnitary(TSHIFT, wires=0),
+                qml.QutritUnitary(TSHIFT, wires=1),
+                qml.QutritUnitary(TADD, wires=[0, 1]),
+            ],
+            obs.diagonalizing_gates(),
+        )
+        res = dev.expval(obs)
+
+        expected = eval_1 * eval_2
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:855]: Will this be true in general? Aren't there states for which the expval of a tensor product is not necessarily the product of expvals?

File: tests/devices/test_default_qutrit.py [hunk 13]
```diff
@@ -797,15 +830,171 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
 
-# TODO: Add tests for tensor non-parametrized observables
 class TestTensorVar:
-    pass
+    """Tests for variance of tensor observables"""
+
+    @pytest.mark.parametrize("index_1", list(range(1, 9)))
+    @pytest.mark.parametrize("index_2", list(range(1, 9)))
+    def test_gell_mann_tensor(self, index_1, index_2, tol):
+        """Test that the variance of tensor Gell-Mann observables is correct"""
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:865]:
```suggestion
        """Test that the variance of tensor products of Gell-Mann observables is correct"""
```

File: tests/devices/test_default_qutrit.py [hunk 14]
```diff
@@ -797,15 +830,171 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
+
+        state = np.array([[0, 0, 0, 0, 0, 1, 1, 0, 0]]) / math.sqrt(2)
+        obs_mat = np.kron(GELL_MANN[index_1 - 1], GELL_MANN[index_2 - 1])
+
+        expected = (
+            state.conj() @ obs_mat @ obs_mat @ state.T - (state.conj() @ obs_mat @ state.T) ** 2
+        )
+        assert np.isclose(res, expected[0], atol=tol, rtol=0)
+
+    def test_hermitian(self, tol):
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:888]: Missing docstring for this test; was it just missed from a previous PR when Hermitian was added?

File: tests/devices/test_default_qutrit.py [hunk 15]
```diff
@@ -797,15 +830,171 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
+        dev = qml.device("default.qutrit", wires=3)
+
+        A1 = np.array([[3, 0, 0], [0, -3, 0], [0, 0, -2]])
+
+        A = np.array([[1, 0, 0], [0, -1, 0], [0, 0, 2]])
+        B = np.array([[4, 0, 0], [0, -2, 0], [0, 0, 1]])
+        A2 = np.kron(A, B)
+
+        obs = qml.THermitian(A1, wires=[0]) @ qml.THermitian(A2, wires=[1, 2])
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:897]: I suggest to also do a test where you tensor together a `GellMannObs` with a `THermitian` obs (both expval and variance).

File: tests/devices/test_default_qutrit.py [hunk 16]
```diff
@@ -797,15 +830,171 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
+        assert np.isclose(res, expected[0], atol=tol, rtol=0)
+
+    def test_hermitian(self, tol):
+        dev = qml.device("default.qutrit", wires=3)
+
+        A1 = np.array([[3, 0, 0], [0, -3, 0], [0, 0, -2]])
+
+        A = np.array([[1, 0, 0], [0, -1, 0], [0, 0, 2]])
+        B = np.array([[4, 0, 0], [0, -2, 0], [0, 0, 1]])
+        A2 = np.kron(A, B)
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:895]: Add in a non-diagonal test case here too, ideally with some complex values.

File: tests/devices/test_default_qutrit.py [hunk 17]
```diff
@@ -797,15 +830,171 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
 
 # TODO: Add tests for tensor non-parametrized observables
 class TestTensorSample:
-    pass
+    """Test tensor samples"""
+
+    @pytest.mark.parametrize("index_1", list(range(1, 9)))
+    @pytest.mark.parametrize("index_2", list(range(1, 9)))
+    def test_gell_mann_obs(self, index_1, index_2, tol_stochastic):
+        """Test that the tensor proudct involving Gell-Mann observables works correctly"""
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:921]:
```suggestion
        """Test that sampling tensor products  involving Gell-Mann observables works correctly"""
```

File: tests/devices/test_default_qutrit.py [hunk 18]
```diff
@@ -797,15 +830,171 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
+
+        var = (s1**2) @ p - (s1 @ p) ** 2
+        expected = (
+            state.conj().T @ obs_mat @ obs_mat @ state - (state.conj().T @ obs_mat @ state) ** 2
+        )
+        assert np.allclose(var, expected, atol=tol_stochastic, rtol=0)
+
+    @pytest.mark.parametrize("index", list(range(1, 9)))
+    def test_hermitian(self, index, tol_stochastic):
+        """Tests that tensor product of hermitian obervable with another observable works
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:959]:
```suggestion
        """Tests that sampling on a tensor product of Hermitian observables with another observable works
```

File: tests/devices/test_default_qutrit.py [hunk 19]
```diff
@@ -797,15 +830,171 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
+        correctly"""
+
+        dev = qml.device("default.qutrit", wires=3, shots=int(1e6))
+
+        A = np.array([[1, 0, 0], [0, -1, 0], [0, 0, 2]])
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:964]: Recommend to try with a non-diagonal / complex case as well. Maybe just add as a new parameter to the test?

File: pennylane/ops/qutrit/observables.py [hunk 20]
```diff
@@ -133,3 +146,136 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+            immediately pushed into the Operator queue (optional)
+        id (str or None): String representing the operation (optional)
+    """
+    num_wires = 1
+    num_params = 1
+    """int: Number of trainable parameters the operator depends on"""
+
+    def __init__(self, index, wires, do_queue=True, id=None):
+        if not isinstance(index, int) or index < 1 or index > 8:
+            raise ValueError("The index must be an integer between 1 and 8 inclusive.")
```
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:188]:
```suggestion
            raise ValueError("The index of a Gell-Mann observable must be an integer between 1 and 8 inclusive.")
```
Just so it's a bit clearer if the error message pops up; the corresponding test case would need to be updated as well.

File: pennylane/ops/qutrit/observables.py [hunk 21]
```diff
@@ -133,3 +146,136 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+    * Number of parameters: 1
+    * Gradient recipe: None
+
+    Args:
+        index (int): The index of the Gell-Mann matrix to be used. Must be between 1
+            and 8 inclusive
+        wires (Sequence[int] or int): the wire(s) the observable acts on
+        do_queue (bool): Indicates whether the operator should be
+            immediately pushed into the Operator queue (optional)
+        id (str or None): String representing the operation (optional)
```
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:180]: I think it'd be good to have a usage example here that shows a qutrit circuit where a Gell-Mann obs is measured, and including both the numerical output as well as the text-based circuit diagram.

File: tests/ops/qutrit/test_qutrit_observables.py [hunk 22]
```diff
@@ -313,3 +314,63 @@ def test_matrix_representation(self, tol):
...
+            assert diag_gates[0].__class__ == qml.QutritUnitary
+            mat = diag_gates[0].matrix()
+            U = np.dot(np.eye(3), mat)
+
+        res = U @ A @ U.conj().T
+        expected = np.diag(eigs)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("index", [0, 9, 1.0, 2.5, "c"])
+    def test_index_error(self, index):
```
Comment (glassnotes) [inline tests/ops/qutrit/test_qutrit_observables.py:356]: Please add docstrings to this test and the next two as well.

**Review by glassnotes — CHANGES_REQUESTED** [review]
Nice work @mudit2812 ! :rocket:

Mostly just suggestions for tests, but there are a couple of important points I wanted to note at a high level:
 - what about renaming `GellMannObs` to just `GellMann`?
 - the `index` parameter is currently being passed like normal trainable parameters usually are; I think it should be specified as a keyword argument after `wires` to be consistent with other operations that have non-trainable parameters

File: pennylane/ops/qutrit/observables.py (hunk 1 above)
Comment (mudit2812) [reply pennylane/ops/qutrit/observables.py:17]: This is because PL numpy wraps arrays inside tensor objects, so if someone calls the `compute_matrix` method, it would be wrapped as a tensor with `requires_grad=True`. This doesn't really break anything, but since this is meant to be a non-parametrized observable, it's matrix shouldn't be differentiable, and any PL code using the observable, such as the `DefaultQutrit` measurement functions, would cast the matrix to match the dtype of the device, so using PL numpy is really unnecessary.

File: pennylane/ops/qutrit/observables.py (hunk 3 above)
Comment (mudit2812) [reply pennylane/ops/qutrit/observables.py:34]: The ordering follows the same comvention as the Pauli matrices. I'll add this as a comment to the file as well, but essentially, matrices 0, 3, and 5 are qutrit analogs of Pauli X with each matrix spanning a different subspace, matrices 1, 4, and 6 are qutrit analogs of Pauli Y, and matrices 2 and 7 are analogs of Pauli Z. There are only 2 matrices to generalize Pauli Z because only 8 matrices are needed for 3 dimensions to create a complete basis.

File: pennylane/ops/qutrit/observables.py (hunk 4 above)
Comment (mudit2812) [reply pennylane/ops/qutrit/observables.py:281]: That's a good idea. I will add this.

File: doc/releases/changelog-dev.md (hunk 7 above)
Comment (mudit2812) [reply doc/releases/changelog-dev.md:76]: I think that makes sense. My reasoning was mostly arbitrary, and this would make it more consistent with the rest of the operators.

File: tests/devices/test_default_qutrit.py (hunk 6 above)
Comment (mudit2812) [reply tests/devices/test_default_qutrit.py:405]: Yep, this is expected behaviour.

File: pennylane/ops/qutrit/observables.py (hunk 1 above)
Comment (mudit2812) [reply pennylane/ops/qutrit/observables.py:17]: This is because PL numpy wraps arrays inside tensor objects, so if someone calls the `compute_matrix` method, it would be wrapped as a tensor with `requires_grad=True`. This doesn't really break anything, but since this is meant to be a non-parametrized observable, it's matrix shouldn't be differentiable, and any PL code using the observable, such as the `DefaultQutrit` measurement functions, would cast the matrix to match the dtype of the device, so using PL numpy is really unnecessary.

File: tests/devices/test_default_qutrit.py (hunk 12 above)
Comment (mudit2812) [reply tests/devices/test_default_qutrit.py:855]: I think that this test passes due to the fact that the measured state doesn't have any entanglement. I'll try out a circuit with entanglement and test whether this still works, which it likely wont, in which case I'll edit the expected result to work with the circuit.

File: pennylane/ops/qutrit/observables.py (hunk 9 above)
Comment (mudit2812) [reply pennylane/ops/qutrit/observables.py:161]: I fixed the alignment issue. Let me know if it looks better now.

File: pennylane/ops/qutrit/observables.py (hunk 1 above)
Comment (glassnotes) [reply pennylane/ops/qutrit/observables.py:17]: Ah good point! Thanks for clarifying. Yeah I see now in the qubit non-parametric ops and observable file this is done as well.

File: pennylane/ops/qutrit/observables.py (hunk 9 above)
Comment (glassnotes) [reply pennylane/ops/qutrit/observables.py:161]: Beautiful :art:

File: pennylane/ops/qutrit/observables.py [hunk 23]
```diff
@@ -133,3 +149,201 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+    ...     qml.TShift(wires=0)
+    ...     qml.TClock(wires=0)
+    ...     qml.TShift(wires=1)
+    ...     qml.TAdd(wires=[0, 1])
+    ...     return qml.expval(qml.GellMann(wires=0, index=1))
+    >>> print(test_qnode())
+    0.0
+    >>> print(qml.draw(test_qnode)())
+    0: ──TShift──TClock─╭●────┤  <GellMann(1)>
+    1: ──TShift─────────╰TAdd─┤
```
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:199]: Nice example -  I am looking at it in the rendered docs, it's so nice to see things come together :100:

File: pennylane/ops/qutrit/observables.py [hunk 24]
```diff
@@ -133,3 +149,201 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+
+    def __init__(self, wires, index=1, do_queue=True, id=None):
+        if not isinstance(index, int) or index < 1 or index > 8:
+            raise ValueError(
+                "The index of a Gell-Mann observable must be an integer between 1 and 8 inclusive."
+            )
+
+        self._hyperparameters = {
+            "index": index,
+        }
```
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:214]:
```suggestion
        self.hyperparameters["index"] = index
```
Since `hyperparameters` is a property of the base class, it can be access directly like this (this is consistent as well with how it gets used in other non-parametric qubit ops).

File: pennylane/ops/qutrit/observables.py [hunk 25]
```diff
@@ -133,3 +149,201 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+
+        super().__init__(wires=wires, do_queue=do_queue, id=id)
+
+    def label(self, decimals=None, base_label=None, cache=None):
+        return base_label or "GellMann(" + str(self._hyperparameters["index"]) + ")"
```
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:219]:
```suggestion
        return base_label or "GellMann(" + str(self.hyperparameters["index"]) + ")"
```

File: pennylane/ops/qutrit/observables.py [hunk 26]
```diff
@@ -133,3 +149,201 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+
+    _eigvecs = {
```
Comment (glassnotes) [inline pennylane/ops/qutrit/observables.py:221]: Rather than making these part of the class, what about defining the `eigendecomposition` property as is done in `Hermitian`? I'm not up-to-date on what's been going on with all the operator arithmetic and new properties, though. @albi3ro could you please provide some guidance here?

File: tests/devices/test_default_qutrit.py [hunk 27]
```diff
@@ -797,15 +855,206 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
+    def test_gell_mann_tensor(self, index_1, index_2, tol):
+        """Test that the expectation value of the tensor product of two Gell-Mann observables is
+        correct"""
+        dev = qml.device("default.qutrit", wires=2)
+        obs = qml.GellMann(wires=0, index=index_1) @ qml.GellMann(wires=1, index=index_2)
+
+        dev.apply(
+            [
+                qml.QutritUnitary(U_thadamard_01, wires=0),
+                qml.QutritUnitary(TADD, wires=[0, 1]),
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:869]: Thanks for updating this test - this looks much safer!

File: tests/devices/test_default_qutrit.py [hunk 28]
```diff
@@ -797,15 +855,206 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
+        state = np.array([[0, 0, 0, 0, 0, 1, 1, 0, 0]]) / math.sqrt(2)
+        obs_mat = np.kron(GELL_MANN[index - 1], A)
+
+        expected = (
+            state.conj() @ obs_mat @ obs_mat @ state.T - (state.conj() @ obs_mat @ state.T) ** 2
+        )
+        assert np.isclose(res, expected[0], atol=tol, rtol=0)
+
+    def test_hermitian(self, tol):
+        """Test that the variance of tensor product of two Hermitian matrices behaves correctly"""
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:940]:
```suggestion
        """Test that the variance of a tensor product of two Hermitian matrices behaves correctly"""
```

File: tests/devices/test_default_qutrit.py [hunk 29]
```diff
@@ -797,15 +855,206 @@ def test_hermitian_two_wires_identity_expectation(self, tol):
...
 
 # TODO: Add tests for tensor non-parametrized observables
 class TestTensorSample:
-    pass
+    """Test tensor samples"""
+
+    @pytest.mark.parametrize("index_1", list(range(1, 9)))
+    @pytest.mark.parametrize("index_2", list(range(1, 9)))
+    def test_gell_mann_obs(self, index_1, index_2, tol_stochastic):
+        """Test that sampling tensor products  involving Gell-Mann observables works correctly"""
```
Comment (glassnotes) [inline tests/devices/test_default_qutrit.py:981]:
```suggestion
        """Test that sampling tensor products involving Gell-Mann observables works correctly"""
```

**Review by glassnotes — CHANGES_REQUESTED** [review]
Thanks @mudit2812 for the updates and clarifications! Just a few remaining comments.

File: pennylane/ops/qutrit/observables.py (hunk 26 above)
Comment (mudit2812) [reply pennylane/ops/qutrit/observables.py:219]: `Hermitian.eigendecomposition` internally uses a dictionary which uses the flattened Hermitian matrix as a key to access the eigenvectors/values. As far as I can tell, the purpose of it is to ensure that if there are multiple instances of `qml.Hermitian`, the eigenvectors/values don't need to be calculated with `np.eigh` every time the observable is used in a circuit or the user requests eigenvalues/vectors, but can rather be accessed directly from the dictionary. What I've done achieves the same for `qml.GellMann`, with the key difference being that there is no reliance on `np.eigh` at any point, as the eigenvectors are hardcoded, and the eigenvalues are too trivial to require a property analogous to `eigendecomposition`.

Let me know if you think that it's still worthwhile going with a structure similar to `Hermitian.eigendecomposition`, and I can update the class in that case.

Comment (Jaybsoni) [reply pennylane/ops/qutrit/observables.py:219]: Actually we are moving away from using the matrix as the hash. Instead we have well defined hash functions for each operator, we use that in combination with the wire order to hash the Eigen decompositions. 

In this case I agree with @mudit2812, you probably don't need to use `np.eigh` to save the Eigen decomposition as the matrices are well known (for the most part) before hand.

**Comment by Jaybsoni** [comment]
Hi @mudit2812, I apologize for the delay! I will leave a review for you later tonight. Thanks,

File: pennylane/ops/qutrit/observables.py [hunk 30]
```diff
@@ -133,3 +149,199 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+            list[.Operator]: list of diagonalizing gates
+
+        **Example**
+
+        >>> qml.GellMann.compute_diagonalizing_gates(wires=0, index=4)
+        [QutritUnitary(array([[ 0.70710678-0.j,  0.        -0.j,  0.70710678-0.j],
+               [ 0.70710678-0.j,  0.        -0.j, -0.70710678-0.j],
+               [ 0.        -0.j,  1.        -0.j,  0.        -0.j]]), wires=[0])]
+        """
+        if index in {3, 8}:
```
Comment (Jaybsoni) [inline pennylane/ops/qutrit/observables.py:343]: Any particular reason for using a set here? 
```suggestion
        if index in (3, 8):
```

File: pennylane/ops/qutrit/observables.py [hunk 31]
```diff
@@ -133,3 +149,199 @@ def compute_diagonalizing_gates(eigenvectors, wires):  # pylint: disable=argumen
...
+            return []
```
Comment (Jaybsoni) [inline pennylane/ops/qutrit/observables.py:344]: Adding a comment for devs, 
```suggestion
            return []  # already diagonal
```

**Review by Jaybsoni — APPROVED** [review]
Looks great to me 💯! just a few minor comments, nothing blocking. 

One question I had, the Pauli ops are both observables and operations. It seems that some of the GM matrices are also unitary, would it make sense to make them operations as well?

**Comment by mudit2812** [comment]
> One question I had, the Pauli ops are both observables and operations. It seems that some of the GM matrices are also unitary, would it make sense to make them operations as well?

I don't think any of the GM matrices are unitary. I'll double check and get back.

**Review by glassnotes — APPROVED** [review]
Thanks @mudit2812 for incorporating all the changes! I added `GellMann` to the docs, and once the build finishes I will merge.
